core/tx_pool.go

```diff
@@ -675,14 +718,282 @@ func (pool *TxPool) validateTx(tx *types.Transaction, local bool) error {
+		}
+		if from != stkMsg.ValidatorAddress {
+			return errors.WithMessagef(ErrInvalidSender, "staking transaction sender is %s", b32)
+		}
+		return pool.validateCreateValidatorMsg(stkMsg)
+	case staking.DirectiveEditValidator:
+		msg, err := staking.RLPDecodeStakeMsg(tx.Data(), staking.DirectiveEditValidator)
+		if err != nil {
+			return err
+		}
+		stkMsg, ok := msg.(*staking.EditValidator)
+		if !ok {
+			return ErrInvalidMsgForStakingDirective
+		}
+		if from != stkMsg.ValidatorAddress {
+			return errors.WithMessagef(ErrInvalidSender, "staking transaction sender is %s", b32)
+		}
+		return pool.validateEditValidatorMsg(stkMsg)
+	case staking.DirectiveDelegate:
+		msg, err := staking.RLPDecodeStakeMsg(tx.Data(), staking.DirectiveDelegate)
+		if err != nil {
+			return err
+		}
+		stkMsg, ok := msg.(*staking.Delegate)
+		if !ok {
+			return ErrInvalidMsgForStakingDirective
+		}
+		if from != stkMsg.DelegatorAddress {
+			return errors.WithMessagef(ErrInvalidSender, "staking transaction sender is %s", b32)
+		}
+		return pool.validateDelegateMsg(stkMsg)
+	case staking.DirectiveUndelegate:
+		msg, err := staking.RLPDecodeStakeMsg(tx.Data(), staking.DirectiveUndelegate)
+		if err != nil {
+			return err
+		}
+		stkMsg, ok := msg.(*staking.Undelegate)
+		if !ok {
+			return ErrInvalidMsgForStakingDirective
+		}
+		if from != stkMsg.DelegatorAddress {
+			return errors.WithMessagef(ErrInvalidSender, "staking transaction sender is %s", b32)
+		}
+		return pool.validateUndelegateMsg(stkMsg)
+	case staking.DirectiveCollectRewards:
+		msg, err := staking.RLPDecodeStakeMsg(tx.Data(), staking.DirectiveUndelegate)
+		if err != nil {
+			return err
+		}
+		stkMsg, ok := msg.(*staking.CollectRewards)
+		if !ok {
+			return ErrInvalidMsgForStakingDirective
+		}
+		if from != stkMsg.DelegatorAddress {
+			return errors.WithMessagef(ErrInvalidSender, "staking transaction sender is %s", b32)
+		}
+		return pool.validateCollectRewardsMsg(stkMsg)
+	default:
+		return ErrUnknownStakingDirective
+	}
+}
+
+func (pool *TxPool) validateCreateValidatorMsg(stkMsg *staking.CreateValidator) error {
+	if pool.currentState.IsValidator(stkMsg.ValidatorAddress) {
+		return errValidatorExist
+	}
+
+	// create temp validator to do sanity check
+	validatorWrapper, err := staking.CreateValidatorFromNewMsg(stkMsg, pool.chain.CurrentBlock().Number())
+	if err != nil {
+		return errors.WithMessage(err, "from verification of create validator message")
+	}
+	w := staking.ValidatorWrapper{}
+	w.Validator = *validatorWrapper
+	w.Delegations = []staking.Delegation{ // Create validators implies self delegation.
+		staking.NewDelegation(validatorWrapper.Address, stkMsg.Amount),
+	}
+
+	return w.SanityCheck()
+}
+
+func (pool *TxPool) validateEditValidatorMsg(stkMsg *staking.EditValidator) error {
+	if !pool.currentState.IsValidator(stkMsg.ValidatorAddress) {
+		return errValidatorNotExist
+	}
+
+	validatorWrapper := pool.currentState.GetStakingInfo(stkMsg.ValidatorAddress)
```

Comment:
same here. Should refactor all these logic into a single function that can be called directly.